src/main.cpp

```diff
@@ -0,0 +1,233 @@
+      read_buf[i] = 0x00;
+    }
+    const int rxBytes = uart_read_bytes(UART_NUM_1, read_buf, RX_BUF_SIZE, DELAY_MS / portTICK_PERIOD_MS);
+
+    if (rxBytes <= 0) {
+      ESP_LOGI(RX_TASK_TAG, "No bytes, received\n");
+      continue;
+    }
+    ESP_LOGI(RX_TASK_TAG, "Motor cmd : %f\t%f", motor_cmd_1, motor_cmd_2);
+    // ESP_LOGI(RX_TASK_TAG, "Read %d bytes: '%s'", rxBytes, read_buf);
+
+
+    memcpy(&motor_cmd_1, read_buf + 3, sizeof(float));
+    memcpy(&motor_cmd_2, read_buf + 7, sizeof(float));
+
+    set_torque(motor_cmd_1, motor1);
+    set_torque(motor_cmd_2, motor2);
+
+  }
+}
+
+
+static void tx_task(void* arg) {
+  // read imu_value and send them
+  static const char* TX_TASK_TAG = "TX_TASK";
+  esp_log_level_set(TX_TASK_TAG, ESP_LOG_INFO);
+  IMU imu = IMU(0x68);
+  imu.init();
+  imu.subscribe("imu", imu_cb);
+  imu.start(100);
+
+
+  // uart 
+  char send_buf[TX_BUFFER_SIZE];
+
+  while (true) {
+    vTaskDelay(DELAY_MS / portTICK_PERIOD_MS);
+
+    // clear  biffer
+    for (size_t i = 0; i < TX_BUFFER_SIZE; i++) {
+      send_buf[i] = 0x00;
+    }
+
+
+    ESP_LOGE("IMU", "Gravity: %f %f %f", imu_data.gravity.x, imu_data.gravity.y,
+      imu_data.gravity.z);
+    ESP_LOGE("IMU", "Linear Acc: %d %d %d", imu_data.linear_acc.x,
+      imu_data.linear_acc.y, imu_data.linear_acc.z);
+    ESP_LOGE("IMU", "Angular Vel: %d %d %d", imu_data.ang_vel.x,
+      imu_data.ang_vel.y, imu_data.ang_vel.z);
+    ESP_LOGE(TX_TASK_TAG, "Orientation: %f %f %f", imu_data.orientation.x,
+      imu_data.orientation.y, imu_data.orientation.z);
+
+    send_buf[0] = 's';  // starting character
+    send_buf[1] = 'f'; // the type of data being sent
+    send_buf[2] = 2; // sending 2 values, roll and angular velocity
+
+    // send values first 4 bytes are for the roll and next 4 bytes are for the angular velocity
+    memcpy(send_buf + 3, &imu_data.orientation.x, sizeof(float)); // TODO : SEND PITCH
+    memcpy(send_buf + 7, &imu_data.orientation.z, sizeof(float));
+    uart_write_bytes(UART_NUM_1, &send_buf, sizeof(send_buf));
+  }
+
+  imu.stop();
+  imu.unsubscribe("imu");
+}
+
+void init_uart(void) {
+
+  const uart_config_t uart_config = {
+      .baud_rate = 115200,
+      .data_bits = UART_DATA_8_BITS,
+      .parity = UART_PARITY_DISABLE,
+      .stop_bits = UART_STOP_BITS_1,
+      .flow_ctrl = UART_HW_FLOWCTRL_DISABLE,
+      .source_clk = UART_SCLK_DEFAULT,
+  };
+  // We won't use a buffer for sending data.
+  uart_driver_install(UART_NUM_1, RX_BUF_SIZE * 2, 0, 0, NULL, 0);
+  uart_param_config(UART_NUM_1, &uart_config);
+  uart_set_pin(UART_NUM_1, TXD_PIN, RXD_PIN, UART_PIN_NO_CHANGE, UART_PIN_NO_CHANGE);
+}
+
+
+void set_torque(float torque, const Motor& motor) {
+  torque = std::max(-STALL_TORQUE, std::min(STALL_TORQUE, torque));
+  gpio_set_level(motor.dir, (torque > 0) ? 1 : 0);
+  float speed = NO_LOAD_SPEED * (std::abs(torque) / STALL_TORQUE);
+  uint32_t duty = (uint32_t)(speed / NO_LOAD_SPEED) * ((1 << DUTY_RES) - 1);
```

Comment:
We can remove the `NO_LOAD_SPEED`

```suggestion
  float speed = (std::abs(torque) / STALL_TORQUE);
  uint32_t duty = (uint32_t)(speed) * ((1 << DUTY_RES) - 1);
```

Also, we're bitshifting 1 by duty resolution?

Reply:
yeah, your shifting 1 the number of resolution we have. In our case it's 8. The result is 11111111. Equivalent to this is 255. Took it from an example, but i can change it to make it clearer.

Reply:
Ahh okay, gotcha. I'd leave it as is and add a comment saying "bitshifting doesn't loop around and will fill lower significance bits by the resolution" or something similar.

Unless I'm confused and that's what bitshifting always does?